x/stakeibc/keeper/msg_server_create_trade_route_test.go

```diff
@@ -0,0 +1,231 @@
+	s.Require().True(found, "trade route should have been created")
+	s.Require().Equal(expectedRoute, actualRoute, "trade route")
+}
+
+// Tests a successful trade route creation
+func (s *KeeperTestSuite) TestCreateTradeRoute_Success() {
+	msg, expectedRoute := s.SetupTestCreateTradeRoute()
+	s.submitCreateTradeRouteAndValidate(msg, expectedRoute)
+}
+
+// Tests creating a trade route that uses the default pool config values
+func (s *KeeperTestSuite) TestCreateTradeRoute_Success_DefaultPoolConfig() {
+	msg, expectedRoute := s.SetupTestCreateTradeRoute()
+
+	// Update the message and remove some trade config parameters
+	// so that the defaults are used
+	msg.MaxSwapAmount = sdk.ZeroInt()
+	msg.MaxAllowedSwapLossRate = ""
+
+	expectedRoute.TradeConfig.MaxAllowedSwapLossRate = sdk.MustNewDecFromStr(keeper.DefaultMaxAllowedSwapLossRate)
+	expectedRoute.TradeConfig.MaxSwapAmount = keeper.DefaultMaxSwapAmount
+
+	s.submitCreateTradeRouteAndValidate(msg, expectedRoute)
+}
+
+// Tests trying to create a route from an invalid authority
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_Authority() {
+	msg, _ := s.SetupTestCreateTradeRoute()
+
+	msg.Authority = "not-gov-address"
+
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "invalid authority")
+}
+
+// Tests creating a duplicate trade route
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_DuplicateTradeRoute() {
+	msg, _ := s.SetupTestCreateTradeRoute()
+
+	// Store down a trade route so the tx hits a duplicate trade route error
+	s.App.StakeibcKeeper.SetTradeRoute(s.Ctx, types.TradeRoute{
+		RewardDenomOnRewardZone: RewardDenom,
+		HostDenomOnHostZone:     HostDenom,
+	})
+
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "Trade route already exists")
+}
+
+// Tests creating a trade route when the host zone or withdrawal address does not exist
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_HostZoneNotRegistered() {
+	msg, _ := s.SetupTestCreateTradeRoute()
+
+	// Remove the host zone withdrawal address and confirm it fails
+	invalidHostZone := s.MustGetHostZone(HostChainId)
+	invalidHostZone.WithdrawalIcaAddress = ""
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, invalidHostZone)
+
+	_, err := s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "withdrawal account not initialized on host zone")
+
+	// Remove the host zone completely and check that that also fails
+	s.App.StakeibcKeeper.RemoveHostZone(s.Ctx, HostChainId)
+
+	_, err = s.GetMsgServer().CreateTradeRoute(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().ErrorContains(err, "host zone not found")
+}
+
+// Tests creating a trade route where the ICA channels cannot be created
+// because the ICA connections do not exist
+func (s *KeeperTestSuite) TestCreateTradeRoute_Failure_ConnectionNotFound() {
```

Comment:
We assume that the withdrawal connection is gotten from the HostZone and always is valid at the time we are making a tradeRoute